build_stats_functions/build_profile/profile.py

```diff
@@ -0,0 +1,239 @@
+# Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+"""Python Code for grabbing stats from build profile to import into BigQuery."""
+import json
+import logging
+import os
+import re
+import gzip
+import tempfile
+from datetime import date
+
+import functions_framework
+from google.cloud import bigquery
+from google.cloud import logging 
+from google.cloud import storage
+
+class UnableToUnzipFile(Exception):
+  """Unable to unzip input file."""
+
+class IncorrectFileTypeError(Exception):
+  """This file is not correctly formatted in JSON."""
+
+
+class MissingTopLevelError(Exception):
+  """The build profile is missing the top-level object."""
+
+
+class IncorrectProfileFormatError(Exception):
+  """This file is not in the correct Build Profile format."""
+
+
+@functions_framework.cloud_event
+def main(cloud_event: any):
+  """Entry point function that is triggered by uploading build profile.
+  Gathers stats from build profile file and sends to BigQuery. Goes through
+  each thread in the file and for each event in the thread extracts relevant
+  info and writes as an object into bigquery
+  Args:
+    cloud_event: the google cloud event that triggered the function
+  """
+  STORAGE_BUCKET = os.environ.get("STORAGE_BUCKET")
+  TABLE_ID = os.environ.get("TABLE_ID")
+  try:
+    client = logging.Client()
+    logger = client.logger("build-profile-testing")
+    logger.log_text("Successfully connected to GCP Logging Client", severity="INFO")
+  except logging.Error:
+    logger.log_text("Unable to connect to GCP Logging client", severity="WARNING")
+    return
+  data = cloud_event.data
+  if "name" not in data:
+    logger.log_text("No filename was found", severity="WARNING")
+    return
+  file_name = data["name"]
+  if not check_path(file_name):
+    logger.log_text(
+        "The current file is not a build profile according to its path", severity="INFO"
+    )
+    return
+  else:
+    logger.log_text("Build profile found", severity="INFO")
+    r = re.search("/[0-9].*", file_name)
+    index = r.span()[0]
+    job = file_name[0:index]
```

Comment:
This should have some error checking, it is pretty brittle in looking how the job's name is formed.

Reply:
So right now the current approach is getting the full file name path which is job_path/build_id/file_name. All the build id's are numbers so I'm having it go through the string and pull everything in the string up until build id because that is then the string name. What kind of error checking should I do?

Reply:
I could always do re.search("[A-Za-z0-9]{8}-[A-Za-z0-9]{4}-[A-Za-z0-9]{4}-[A-Za-z0-9]{4}-[A-Za-z0-9]{12}") which would be the regex matching the occurence of the build_id in the path so would go up until that point in the path to get the full job name